app/controllers/api_controller.rb

```diff
@@ -3,6 +3,52 @@ class ApiController < ApplicationController
 
   private
 
+  ##
+  # Set default request format to xml unless a client requests a specific format,
+  # which can be done via (a) URL suffix and/or (b) HTTP Accept header, where
+  # the URL suffix always takes precedence over the Accept header.
+  def set_default_request_format
```

Comment:
>The edge-case is that if the header exists, and is parseable, but doesn't contain either xml, json, or '/', then request.format isn't set. I'm not sure if that has any consequences though.

That's not really an edge case, leaving request.format as is is intended behaviour and covered by the last Accept header unit test  "text/json is in invalid format, ActionController::UnknownFormat error is expected". If a user requests a single format we don't support at all, we can't fall back to XML here, as it would be violating the respective RFC.

In any case, it would be better to return 406 Not Acceptable instead of HTTP 500. Somehow the Rails framework doesn't handle this properly. In any case, that's outside of anything I control.

>  assume request.format = "xml" on the first line of the function. 

No, don't do that. As I said, it would introduce incorrect behaviour.

> , if you've had enough of refactoring this function already, I understand.

Right. No more refactoring from my side. I needed some code I can still read in a week's time. You're of course free to make it look more Rails like. It would take me more than an hour and the result would still be poor.

Reply:
Wait, I really wanted to add a test case with "`http_accept_format("image/jpeg")`", which behaves in the way like text/json (which is not a valid format). Will to that. The expected outcome would still be HTTP 500 (instead of HTTP 406), and not returning XML.

Reply:
> It would be better to return 406 Not Acceptable instead of HTTP 500. Somehow the Rails framework doesn't handle this properly. but unfortunately that's outside of anything I control.

I've investigated this, and I believe it's because we wrap all api calls in our own `api_call_handle_error`. You might be able to solve this by adding

```
rescue ActionController::UnknownFormat
   head :not_acceptable
```
to the top of the list of handled exceptions at https://github.com/openstreetmap/openstreetmap-website/blob/f7f73322b3abae8648d6462914eea032a93263fd/app/controllers/application_controller.rb#L186-L204

Reply:
Ah, thanks! That's much better, I will change this in a minute.